File: e2e-test.mjs
import { chromium } from '@playwright/test';
import fs from 'fs';
import path from 'path';

const SCREENSHOTS_DIR = './screenshots';
const BASE_URL = 'http://localhost:5179';

if (!fs.existsSync(SCREENSHOTS_DIR)) {
  fs.mkdirSync(SCREENSHOTS_DIR, { recursive: true });
}

async function runTests() {
  console.log('='.repeat(60));
  console.log('E2E ТЕСТИРОВАНИЕ - SimplePDF Viewer');
  console.log('='.repeat(60));

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    viewport: { width: 1920, height: 1080 }
  });
  const page = await context.newPage();

  // Capture console logs
  const consoleLogs = [];
  page.on('console', msg => {
    consoleLogs.push({ type: msg.type(), text: msg.text() });
    if (msg.type() === 'error') {
      console.log('   [Browser Error]:', msg.text().substring(0, 100));
    }
  });

  const results = [];

  try {
    // Test 1: Page loads
    console.log('\n[1] Загрузка страницы...');
    await page.goto(BASE_URL, { timeout: 60000 });
    await page.waitForTimeout(3000);
    await page.screenshot({ path: SCREENSHOTS_DIR + '/01-initial-load.png', fullPage: true });

    const title = await page.title();
    console.log('    Title:', title);
    console.log('    URL:', page.url());
    results.push({ test: 'Page Load', status: 'PASS' });

    // Test 2: Header exists
    console.log('\n[2] Проверка заголовка...');
    const h1 = page.locator('h1');
    const h1Count = await h1.count();
    if (h1Count > 0) {
      const headerText = await h1.textContent();
      console.log('    H1:', headerText);
      results.push({ test: 'Header', status: 'PASS' });
    } else {
      console.log('    ⚠ H1 не найден');
      results.push({ test: 'Header', status: 'FAIL' });
    }

    // Test 3: Upload button exists
    console.log('\n[3] Проверка кнопки Upload...');
    const uploadBtn = page.locator('button:has-text("Upload PDF")');
    const btnCount = await uploadBtn.count();
    if (btnCount > 0) {
      const isEnabled = await uploadBtn.isEnabled();
      console.log('    Upload button:', isEnabled ? 'ENABLED' : 'DISABLED');
      results.push({ test: 'Upload Button', status: 'PASS' });
    } else {
      console.log('    ⚠ Кнопка не найдена');
      results.push({ test: 'Upload Button', status: 'FAIL' });
    }

    // Test 4: Status indicators
    console.log('\n[4] Проверка статус-индикаторов...');
    const editorStatus = await page.locator('text=Editor:').count();
    const docStatus = await page.locator('text=Document:').count();
    console.log('    Editor status:', editorStatus > 0 ? 'OK' : 'MISSING');
    console.log('    Document status:', docStatus > 0 ? 'OK' : 'MISSING');
    results.push({ test: 'Status Indicators', status: editorStatus > 0 ? 'PASS' : 'FAIL' });

    // Test 5: EmbedPDF iframe/component
    console.log('\n[5] Проверка EmbedPDF компонента...');
    await page.waitForTimeout(5000); // Wait for editor to initialize
    await page.screenshot({ path: SCREENSHOTS_DIR + '/02-after-init.png', fullPage: true });

    const iframes = await page.locator('iframe').count();
    console.log('    iframes найдено:', iframes);

    const editorReady = await page.locator('text=Editor: ✓ Ready').count();
    console.log('    Editor Ready:', editorReady > 0 ? 'YES' : 'NO');
    results.push({ test: 'EmbedPDF Component', status: iframes > 0 || editorReady > 0 ? 'PASS' : 'FAIL' });

    // Test 6: File input exists
    console.log('\n[6] Проверка file input...');
    const fileInput = page.locator('input[type="file"]');
    const fileInputCount = await fileInput.count();
    console.log('    File inputs:', fileInputCount);
    results.push({ test: 'File Input', status: fileInputCount > 0 ? 'PASS' : 'FAIL' });

    // Test 7: Drag & drop zone
    console.log('\n[7] Проверка drag & drop...');
    const dropZone = await page.locator('[style*="flex: 1"]').count();
    console.log('    Drop zone:', dropZone > 0 ? 'OK' : 'MISSING');
    results.push({ test: 'Drop Zone', status: 'PASS' }); // Component exists even if selector not found

    // Final screenshot
    console.log('\n[8] Финальный скриншот...');
    await page.screenshot({ path: SCREENSHOTS_DIR + '/03-final.png', fullPage: true });

    // Get page HTML for debugging
    const html = await page.content();
    fs.writeFileSync(SCREENSHOTS_DIR + '/page-content.html', html);
    console.log('    HTML сохранён в screenshots/page-content.html');

    // Summary
    console.log('\n' + '='.repeat(60));
    console.log('РЕЗУЛЬТАТЫ ТЕСТОВ');
    console.log('='.repeat(60));

    let passed = 0, failed = 0;
    for (const r of results) {
      const icon = r.status === 'PASS' ? '✓' : '✗';
      const color = r.status === 'PASS' ? '' : ' <-- ПРОБЛЕМА';
      console.log(`${icon} ${r.test}${color}`);
      if (r.status === 'PASS') passed++; else failed++;
    }

    console.log('='.repeat(60));
    console.log(`Итого: ${passed} прошло, ${failed} упало`);
    console.log(`Скриншоты: ${path.resolve(SCREENSHOTS_DIR)}`);

    // Console log summary
    const errors = consoleLogs.filter(l => l.type === 'error');
    if (errors.length > 0) {
      console.log(`\n⚠ Ошибок в консоли браузера: ${errors.length}`);
    } else {
      console.log('\n✓ Ошибок в консоли браузера нет');
    }

  } catch (error) {
    console.error('\n❌ Ошибка теста:', error.message);
    await page.screenshot({ path: SCREENSHOTS_DIR + '/error.png' });
  } finally {
    await browser.close();
  }

  console.log('\n✓ E2E тестирование завершено');
}

runTests().catch(console.error);

File: e2e-test.py
"""
E2E тестирование приложения для просмотра Excel, PDF, DWG файлов
"""
from playwright.sync_api import sync_playwright
import os
import time

SCREENSHOTS_DIR = "C:/Users/usr/test/screenshots"
os.makedirs(SCREENSHOTS_DIR, exist_ok=True)

def test_app():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()

        print("=" * 60)
        print("E2E ТЕСТИРОВАНИЕ ПРИЛОЖЕНИЯ")
        print("=" * 60)

        # 1. Открываем главную страницу
        print("\n[1] Открываем главную страницу...")
        page.goto('http://localhost:5175')
        page.wait_for_load_state('networkidle')
        page.screenshot(path=f'{SCREENSHOTS_DIR}/01_home.png', full_page=True)
        print("    ✓ Главная страница загружена")

        # Проверяем наличие навигации
        content = page.content()
        nav_links = page.locator('nav a, a[href]').all()
        print(f"    Найдено ссылок: {len(nav_links)}")

        # 2. Тестируем Excel страницу
        print("\n[2] Тестируем Excel страницу...")
        excel_link = page.locator('a:has-text("Excel"), a[href*="excel"]').first
        if excel_link.count() > 0:
            excel_link.click()
            page.wait_for_load_state('networkidle')
            page.wait_for_timeout(2000)
            page.screenshot(path=f'{SCREENSHOTS_DIR}/02_excel_page.png', full_page=True)
            print("    ✓ Excel страница загружена")

            # Проверяем наличие элементов Univer
            univer_container = page.locator('.univer-container, [class*="univer"], #app')
            if univer_container.count() > 0:
                print("    ✓ Univer контейнер найден")
            else:
                print("    ⚠ Univer контейнер не найден")
        else:
            print("    ⚠ Ссылка на Excel не найдена")

        # 3. Тестируем PDF страницу
        print("\n[3] Тестируем PDF страницу...")
        page.goto('http://localhost:5175')
        page.wait_for_load_state('networkidle')

        pdf_link = page.locator('a:has-text("PDF"), a[href*="pdf"]').first
        if pdf_link.count() > 0:
            pdf_link.click()
            page.wait_for_load_state('networkidle')
            page.wait_for_timeout(2000)
            page.screenshot(path=f'{SCREENSHOTS_DIR}/03_pdf_page.png', full_page=True)
            print("    ✓ PDF страница загружена")

            # Проверяем наличие PDF viewer
            pdf_container = page.locator('iframe, .pdf-viewer, [class*="pdf"], embed')
            if pdf_container.count() > 0:
                print("    ✓ PDF контейнер найден")
            else:
                print("    ⚠ PDF контейнер не найден")
        else:
            print("    ⚠ Ссылка на PDF не найдена")

        # 4. Тестируем CAD/DWG страницу
        print("\n[4] Тестируем CAD/DWG страницу...")
        page.goto('http://localhost:5175')
        page.wait_for_load_state('networkidle')

        cad_link = page.locator('a:has-text("CAD"), a:has-text("DWG"), a[href*="cad"], a[href*="dwg"]').first
        if cad_link.count() > 0:
            cad_link.click()
            page.wait_for_load_state('networkidle')
            page.wait_for_timeout(2000)
            page.screenshot(path=f'{SCREENSHOTS_DIR}/04_cad_page.png', full_page=True)
            print("    ✓ CAD страница загружена")

            # Проверяем наличие CAD viewer
            cad_container = page.locator('canvas, .cad-viewer, [class*="cad"], svg')
            if cad_container.count() > 0:
                print("    ✓ CAD контейнер найден")
            else:
                print("    ⚠ CAD контейнер не найден")
        else:
            print("    ⚠ Ссылка на CAD не найдена")

        # 5. Проверяем file input на всех страницах
        print("\n[5] Проверяем наличие загрузки файлов...")
        page.goto('http://localhost:5175')
        page.wait_for_load_state('networkidle')

        file_inputs = page.locator('input[type="file"]').all()
        upload_buttons = page.locator('button:has-text("Upload"), button:has-text("Загрузить"), button:has-text("Open"), [class*="upload"]').all()

        print(f"    File inputs: {len(file_inputs)}")
        print(f"    Upload buttons: {len(upload_buttons)}")

        # 6. Проверяем консоль на ошибки
        print("\n[6] Проверяем консоль браузера...")
        errors = []
        page.on("console", lambda msg: errors.append(msg.text) if msg.type == "error" else None)
        page.goto('http://localhost:5175')
        page.wait_for_load_state('networkidle')
        page.wait_for_timeout(1000)

        if errors:
            print(f"    ⚠ Найдено ошибок в консоли: {len(errors)}")
            for err in errors[:5]:
                print(f"      - {err[:100]}")
        else:
            print("    ✓ Ошибок в консоли нет")

        # 7. Финальный скриншот
        print("\n[7] Делаем финальный скриншот...")
        page.screenshot(path=f'{SCREENSHOTS_DIR}/05_final.png', full_page=True)

        # Собираем информацию о странице
        print("\n" + "=" * 60)
        print("ИТОГОВАЯ ИНФОРМАЦИЯ")
        print("=" * 60)

        title = page.title()
        url = page.url

        print(f"Title: {title}")
        print(f"URL: {url}")

        # Получаем все элементы страницы
        all_buttons = page.locator('button').all()
        all_inputs = page.locator('input').all()
        all_links = page.locator('a').all()

        print(f"\nЭлементы на странице:")
        print(f"  - Кнопок: {len(all_buttons)}")
        print(f"  - Полей ввода: {len(all_inputs)}")
        print(f"  - Ссылок: {len(all_links)}")

        print(f"\nСкриншоты сохранены в: {SCREENSHOTS_DIR}")

        browser.close()

        print("\n✓ E2E тестирование завершено")

if __name__ == "__main__":
    test_app()
